spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -0,0 +1,268 @@
+        table.properties(),
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES,
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT);
+    this.rewriteDataFileSizeInBytes = Math.min(splitSize, targetFileSize);
+
+    this.rewriteDataFileSizeInBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_SIZE,
+        TableProperties.SPLIT_SIZE_DEFAULT);
+    this.splitLookback = PropertyUtil.propertyAsInt(
+        table.properties(),
+        TableProperties.SPLIT_LOOKBACK,
+        TableProperties.SPLIT_LOOKBACK_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      this.fileIO = new HadoopFileIO(conf::value);
+    } else {
+      this.fileIO = table.io();
+    }
+    this.encryptionManager = table.encryption();
+  }
+
+  @Override
+  protected RewriteDataFilesAction self() {
+    return this;
+  }
+
+  @Override
+  protected Table table() {
+    return table;
+  }
+
+  /**
+   * Pass a PartitionSepc id to specify which PartitionSpec should be used in DataFile rewrite
+   *
+   * @param specId PartitionSpec id to rewrite
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction outputSpecId(int specId) {
+    Preconditions.checkArgument(table.specs().containsKey(specId), "Invalid spec id %d", specId);
+    this.spec = table.specs().get(specId);
+    return this;
+  }
+
+  /**
+   * Specify the rewrite data file size in bytes
+   *
+   * @param rewriteDataFileSize size of rewrite data file
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction rewriteDataFileSizeInBytes(long rewriteDataFileSize) {
+    Preconditions.checkArgument(rewriteDataFileSize > 0L, "Invalid rewrite data file size in bytes %d",
+        rewriteDataFileSize);
+    this.rewriteDataFileSizeInBytes = rewriteDataFileSize;
+    return this;
+  }
+
+  /**
+   * Specify the split lookback
+   *
+   * @param lookback lookback number to split
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction splitLookback(int lookback) {
+    Preconditions.checkArgument(lookback > 0L, "Invalid split lookback %d", lookback);
+    this.splitLookback = lookback;
+    return this;
+  }
+
+  /**
+   * Pass a row Expression to filter DataFiles to be rewritten. This is typically used when only rewriting DatFiles
+   * under some partitions.
+   *
+   * @param expr Expression to filter out DataFiles
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction filter(Expression expr) {
+    this.filter = Expressions.and(filter, expr);
+    return this;
+  }
+
+  @Override
+  public RewriteDataFilesActionResult execute() {
+    CloseableIterable<FileScanTask> fileScanTasks = table.newScan()
+        .filter(filter)
+        .planFiles();
```

Comment:
Could this use `ManifestGroup` instead of a scan? The scan will emit a scan event, but this isn't really a table scan.

Also, this should be wrapped in a try-with-resources block so that the `CloseableIterable` is closed.

Reply:
AFAIK, we already called `close` in `groupTasksByPartition`, do we need to call wrap the code here in `try-with-resources` again?

Reply:
Besides, `ManifestGroup` is package private, so it cannot be used here.

Reply:
Yes, we need to close the `CloseableIterable`. That will close all iterators that are still open. The close in `groupTasksByPartition` closes the `CloseableIterator` that was passed in, not the group of iterators created from the iterable. Ideally, there shouldn't be other open iterators, but closing the iterable helps ensure that we catch any that are accidentally missed.